// kepler.gl/dist/components/map/map-popover.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MapPopover = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _class, _temp;

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  ', '\n  font-size: 11px;\n  font-weight: 500;\n  background-color: ', ';\n  color: ', ';\n  z-index: 1001;\n  position: absolute;\n  overflow-x: auto;\n\n  .gutter {\n    height: 6px;\n  }\n\n  table {\n    margin: 2px 12px 12px 12px;\n    width: auto;\n\n    tbody {\n      border-top: transparent;\n      border-bottom: transparent;\n    }\n\n    td {\n      border-color: transparent;\n      padding: 4px;\n      color: ', ';\n    }\n\n    td.row__value {\n      text-align: right;\n      font-weight: 500;\n      color: ', ';\n    }\n  }\n'], ['\n  ', '\n  font-size: 11px;\n  font-weight: 500;\n  background-color: ', ';\n  color: ', ';\n  z-index: 1001;\n  position: absolute;\n  overflow-x: auto;\n\n  .gutter {\n    height: 6px;\n  }\n\n  table {\n    margin: 2px 12px 12px 12px;\n    width: auto;\n\n    tbody {\n      border-top: transparent;\n      border-bottom: transparent;\n    }\n\n    td {\n      border-color: transparent;\n      padding: 4px;\n      color: ', ';\n    }\n\n    td.row__value {\n      text-align: right;\n      font-weight: 500;\n      color: ', ';\n    }\n  }\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  position: absolute;\n  left: 50%;\n  transform: rotate(30deg);\n  top: 10px;\n  color: ', ';\n\n  :hover {\n    cursor: pointer;\n    color: ', ';\n  }\n'], ['\n  position: absolute;\n  left: 50%;\n  transform: rotate(30deg);\n  top: 10px;\n  color: ', ';\n\n  :hover {\n    cursor: pointer;\n    color: ', ';\n  }\n']),
    _templateObject3 = (0, _taggedTemplateLiteral3.default)(['\n  color: ', ';\n  font-size: 12px;\n  letter-spacing: 0.43px;\n  text-transform: capitalize;\n  padding-left: 14px;\n  margin-top: 12px;\n\n  svg {\n    margin-right: 4px;\n  }\n'], ['\n  color: ', ';\n  font-size: 12px;\n  letter-spacing: 0.43px;\n  text-transform: capitalize;\n  padding-left: 14px;\n  margin-top: 12px;\n\n  svg {\n    margin-right: 4px;\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _styledComponents3 = require('../common/styled-components');

var _icons = require('../common/icons');

var _defaultSettings = require('../../constants/default-settings');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MAX_WIDTH = 400;
var MAX_HEIGHT = 600;

var StyledMapPopover = _styledComponents2.default.div(_templateObject, function (props) {
  return props.theme.scrollBar;
}, function (props) {
  return props.theme.panelBackground;
}, function (props) {
  return props.theme.textColor;
}, function (props) {
  return props.theme.textColor;
}, function (props) {
  return props.theme.textColorHl;
});

var StyledPin = _styledComponents2.default.div(_templateObject2, function (props) {
  return props.theme.primaryBtnBgd;
}, function (props) {
  return props.theme.linkBtnColor;
});

var StyledLayerName = _styledComponents3.CenterFlexbox.extend(_templateObject3, function (props) {
  return props.theme.textColorHl;
});

var MapPopover = exports.MapPopover = (_temp = _class = function (_Component) {
  (0, _inherits3.default)(MapPopover, _Component);

  function MapPopover(props) {
    (0, _classCallCheck3.default)(this, MapPopover);

    var _this = (0, _possibleConstructorReturn3.default)(this, (MapPopover.__proto__ || Object.getPrototypeOf(MapPopover)).call(this, props));

    _this.state = {
      isMouseOver: false,
      width: 380,
      height: 160
    };
    return _this;
  }

  (0, _createClass3.default)(MapPopover, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this._setContainerSize();
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      this._setContainerSize();
    }
  }, {
    key: '_setContainerSize',
    value: function _setContainerSize() {
      var node = this.popover;
      if (!node) {
        return;
      }

      var width = Math.min(node.scrollWidth, MAX_WIDTH);
      var height = Math.min(node.scrollHeight, MAX_HEIGHT);

      if (width !== this.state.width || height !== this.state.height) {
        this.setState({ width: width, height: height });
      }
    }
  }, {
    key: '_getPosition',
    value: function _getPosition(x, y) {
      var topOffset = 30;
      var leftOffset = 30;
      var mapState = this.props.mapState;
      var _state = this.state,
          width = _state.width,
          height = _state.height;

      var pos = {};
      if (x + leftOffset + width > mapState.width) {
        pos.right = mapState.width - x + leftOffset;
      } else {
        pos.left = x + leftOffset;
      }

      if (y + topOffset + height > mapState.height) {
        pos.bottom = 10;
      } else {
        pos.top = y + topOffset;
      }

      return pos;
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          x = _props.x,
          y = _props.y,
          isVisible = _props.isVisible,
          data = _props.data,
          layer = _props.layer,
          freezed = _props.freezed,
          fields = _props.fields,
          _props$fieldsToShow = _props.fieldsToShow,
          fieldsToShow = _props$fieldsToShow === undefined ? [] : _props$fieldsToShow;

      var hidden = !isVisible && !this.state.isMouseOver;
      var width = this.state.width;


      if (!data || !layer || !fieldsToShow.length) {
        return null;
      }

      var infoProps = { data: data, layer: layer, fieldsToShow: fieldsToShow, fields: fields };

      var style = Number.isFinite(x) && Number.isFinite(y) ? this._getPosition(x, y) : {};

      return _react2.default.createElement(
        StyledMapPopover,
        {
          innerRef: function innerRef(comp) {
            _this2.popover = comp;
          },
          className: (0, _classnames2.default)('map-popover', { hidden: hidden }),
          style: (0, _extends3.default)({}, style, {
            maxWidth: width
          }),
          onMouseEnter: function onMouseEnter() {
            _this2.setState({ isMouseOver: true });
          },
          onMouseLeave: function onMouseLeave() {
            _this2.setState({ isMouseOver: false });
          }
        },
        freezed ? _react2.default.createElement(
          'div',
          { className: 'map-popover__top' },
          _react2.default.createElement('div', { className: 'gutter' }),
          _react2.default.createElement(
            StyledPin,
            { className: 'popover-pin', onClick: this.props.onClose },
            _react2.default.createElement(_icons.Pin, { height: '16px' })
          )
        ) : null,
        _react2.default.createElement(
          StyledLayerName,
          { className: 'map-popover__layer-name' },
          _react2.default.createElement(_icons.Layers, { height: '12px' }),
          layer.config.label
        ),
        _react2.default.createElement(
          'table',
          { className: 'map-popover__table' },
          layer.isAggregated ? _react2.default.createElement(CellInfo, infoProps) : _react2.default.createElement(EntryInfo, infoProps)
        )
      );
    }
  }]);
  return MapPopover;
}(_react.Component), _class.propTypes = {
  fields: _propTypes2.default.arrayOf(_propTypes2.default.any),
  fieldsToShow: _propTypes2.default.arrayOf(_propTypes2.default.any),
  isVisible: _propTypes2.default.bool,
  layer: _propTypes2.default.object,
  data: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.any), _propTypes2.default.object]),
  freezed: _propTypes2.default.bool,
  x: _propTypes2.default.number,
  y: _propTypes2.default.number,
  onClose: _propTypes2.default.func,
  mapState: _propTypes2.default.object.isRequired
}, _temp);


var Row = function Row(_ref) {
  var name = _ref.name,
      value = _ref.value,
      url = _ref.url;

  // Set 'url' to 'value' if it looks like a url
  if (!url && value && typeof value === 'string' && value.match(/^http/)) {
    url = value;
  }

  var asImg = /<img>/.test(name);
  return _react2.default.createElement(
    'tr',
    { className: 'row', key: name },
    _react2.default.createElement(
      'td',
      { className: 'row__name' },
      name
    ),
    _react2.default.createElement(
      'td',
      { className: 'row__value' },
      asImg ? _react2.default.createElement('img', { src: value }) : url ? _react2.default.createElement(
        'a',
        { target: '_blank', rel: 'noopener noreferrer', href: url },
        value
      ) : value
    )
  );
};

var EntryInfo = function EntryInfo(_ref2) {
  var fieldsToShow = _ref2.fieldsToShow,
      fields = _ref2.fields,
      data = _ref2.data;
  return _react2.default.createElement(
    'tbody',
    null,
    fieldsToShow.map(function (name) {
      return _react2.default.createElement(EntryInfoRow, { key: name, name: name, fields: fields, data: data });
    })
  );
};

var EntryInfoRow = function EntryInfoRow(_ref3) {
  var name = _ref3.name,
      fields = _ref3.fields,
      data = _ref3.data;

  var field = fields.find(function (f) {
    return f.name === name;
  });
  if (!field) {
    return null;
  }

  var valueIdx = field.tableFieldIndex - 1;
  var format = _getCellFormat(field.type);

  return _react2.default.createElement(Row, { name: name, value: format ? format(data[valueIdx]) : data[valueIdx] });
};

var CellInfo = function CellInfo(_ref4) {
  var data = _ref4.data,
      layer = _ref4.layer;
  var _layer$config = layer.config,
      colorField = _layer$config.colorField,
      sizeField = _layer$config.sizeField;


  return _react2.default.createElement(
    'tbody',
    null,
    _react2.default.createElement(Row, { name: 'total points', key: 'count', value: data.points && data.points.length }),
    colorField && layer.visualChannels.color ? _react2.default.createElement(Row, {
      name: layer.getVisualChannelDescription('color').measure,
      key: 'color',
      value: data.colorValue || 'N/A'
    }) : null,
    sizeField && layer.visualChannels.size ? _react2.default.createElement(Row, {
      name: layer.getVisualChannelDescription('size').measure,
      key: 'size',
      value: data.elevationValue || 'N/A'
    }) : null
  );
};

function _getCellFormat(type) {
  return _defaultSettings.FIELD_DISPLAY_FORMAT[type];
}

var MapPopoverFactory = function MapPopoverFactory() {
  return MapPopover;
};
exports.default = MapPopoverFactory;
